src/app/clusters/time-synchronization-server/time-synchronization-server.cpp

```diff
@@ -0,0 +1,980 @@
+    {
+        ChipLogError(Zcl, "Time Synchronization failed to schedule timer.");
+    }
+}
+
+CHIP_ERROR TimeSynchronizationServer::SetUTCTime(EndpointId ep, uint64_t utcTime, GranularityEnum granularity,
+                                                 TimeSourceEnum source)
+{
+    ReturnErrorOnFailure(UpdateUTCTime(utcTime));
+    mGranularity = granularity;
+    if (EMBER_ZCL_STATUS_SUCCESS != TimeSource::Set(ep, source))
+    {
+        ChipLogError(Zcl, "Writing TimeSource failed.");
+        return CHIP_IM_GLOBAL_STATUS(Failure);
+    }
+    return CHIP_NO_ERROR;
+}
+
+CHIP_ERROR TimeSynchronizationServer::GetLocalTime(EndpointId ep, DataModel::Nullable<uint64_t> & localTime)
+{
+    int64_t timeZoneOffset = 0, dstOffset = 0;
+    System::Clock::Microseconds64 utcTime;
+    uint64_t chipEpochTime;
+    VerifyOrReturnError(TimeState::kInvalid != GetUpdatedDSTOffsetState(), CHIP_ERROR_INVALID_TIME);
+    ReturnErrorOnFailure(System::SystemClock().GetClock_RealTime(utcTime));
+    VerifyOrReturnError(UnixEpochToChipEpochMicro(utcTime.count(), chipEpochTime), CHIP_ERROR_INVALID_TIME);
+    if (TimeState::kChanged == GetUpdatedTimeZoneState())
+    {
+        sendTimeZoneStatusEvent(ep);
+    }
+    VerifyOrReturnError(GetTimeZone().size() != 0, CHIP_ERROR_INVALID_TIME);
+    const auto & tzStore = GetTimeZone()[0];
+    timeZoneOffset       = static_cast<int64_t>(tzStore.timeZone.offset);
+    VerifyOrReturnError(GetDSTOffset().size() != 0, CHIP_ERROR_INVALID_TIME);
+    const auto & dst = GetDSTOffset()[0];
+    dstOffset        = static_cast<int64_t>(dst.offset);
+
+    uint64_t usRemainder = chipEpochTime % chip::kMicrosecondsPerSecond;   // microseconds part of chipEpochTime
+    chipEpochTime        = (chipEpochTime / chip::kMicrosecondsPerSecond); // make it safe to cast to int64 by converting to seconds
+
+    uint64_t localTimeSec = static_cast<uint64_t>(static_cast<int64_t>(chipEpochTime) + timeZoneOffset + dstOffset);
+    localTime.SetNonNull((localTimeSec * chip::kMicrosecondsPerSecond) + usRemainder);
+    return CHIP_NO_ERROR;
+}
+
+TimeState TimeSynchronizationServer::GetUpdatedTimeZoneState()
+{
+    System::Clock::Microseconds64 utcTime;
+    auto tzList          = GetTimeZone();
+    size_t activeTzIndex = 0;
+    uint64_t chipEpochTime;
+
+    VerifyOrReturnValue(System::SystemClock().GetClock_RealTime(utcTime) == CHIP_NO_ERROR, TimeState::kInvalid);
+    VerifyOrReturnValue(tzList.size() != 0, TimeState::kInvalid);
+    VerifyOrReturnValue(UnixEpochToChipEpochMicro(utcTime.count(), chipEpochTime), TimeState::kInvalid);
+
+    for (size_t i = 0; i < tzList.size(); i++)
+    {
+        auto & tz = tzList[i].timeZone;
+        if (tz.validAt != 0 && tz.validAt <= chipEpochTime)
+        {
+            tz.validAt    = 0;
+            activeTzIndex = i;
+        }
+    }
+    if (activeTzIndex != 0)
+    {
+        mTimeZoneObj.size    = tzList.size() - activeTzIndex;
```

Comment:
Why is this size change here?  If the bits below succeed, LoadTimeZone will change the size as needed, right?  If they don't succeed, this size change is possibly chopping off the now-valid timezone at the end...

Followup is OK for removing this line or documenting why it's here.

Reply:
Tracked in https://github.com/project-chip/connectedhomeip/issues/27575, resolving this comment, follow up in issue.